packages/server/src/ws-connection.ts

```diff
@@ -0,0 +1,184 @@
+    // console.log(`Client ${clientId} disconnected`)
+  })
+
+  function handleSubKey(key: string) {
+    // console.log(`${clientId} sub to key ${key}`)
+    async function send() {
+      const sender = clientSenders.get(clientId)
+      if (!sender) {
+        return
+      }
+      const model = getModel(key) as ValueModel<unknown> | undefined
+      const val = typeof model === 'function' ? model() : model?.get()
+      // console.log('sending', model, key, val)
+      sender({
+        $: 'up',
+        key,
+        val,
+      })
+    }
+    const handlers =
+      clientSubscribers.get(key) ||
+      (() => {
+        const handlers = new Map<string, () => void>()
+        clientSubscribers.set(key, handlers)
+        const model = getModel(key)
+        if (model && typeof model === 'object') {
+          model.subscribe(() => {
+            handlers.forEach((handler) => handler())
+          })
+        }
+
+        return handlers
+      })()
+    handlers.set(clientId, send)
+    send()
+  }
+
+  function handleUnsubKey(key: string) {
+    const handlers = clientSubscribers.get(key)
+    handlers?.delete(clientId)
+  }
+
+  function handleSub({ keys }: z.infer<typeof serverSubscribeMessageSchema>) {
+    keys.forEach(handleSubKey)
+  }
+
+  function handleUnsub({ keys }: z.infer<typeof serverUnsubscribeMessageSchema>) {
+    keys.forEach(handleUnsubKey)
+  }
+
+  async function handleEvt({ key, event }: z.infer<typeof serverEventMessageSchema>) {
+    // console.log('handleEvt', key, event)
+    const { target, payload } = event
+    const { path } = target
+    try {
+      const [storeName, ...lookupPath] = path
+      if (!storeName) throw new Error('Missing store name in event path')
+      if (typeof storeName !== 'string') throw new Error('Store name must be a string')
+      const model = findModel(models, storeName.split('/'))
+      if (!model) throw new Error(`Model not found for store name: ${storeName}`)
+      const modelState = getModelState(model)
+      const value = lookupValue(modelState, lookupPath)
+
+      if (!isServerEventDataState(value)) {
+        throw new Error(`Missing event handler on the server for event: ${JSON.stringify(event)}`)
+      }
+      let res = await value.handler(...payload)
+      if (!isResponseDataState(res)) {
+        res = response(res ?? null)
+      }
+      clientSenders.get(clientId)?.({
+        $: 'evt-res',
+        key,
+        res,
+      })
+    } catch (error: any) {
+      clientSenders.get(clientId)?.({
+        $: 'evt-res',
+        key,
+        res: response(error).status(500),
+      })
+      return
+    }
+  }
+
+  // console.log('connected client', clientId)
+  ws.on('message', (messageData) => {
+    let messageUnvalidated, message
+    try {
+      messageUnvalidated = JSON.parse(messageData.toString())
+      // console.log(messageUnvalidated)
+    } catch (e) {
+      throw new Error('Failed to parse JSON from client ' + clientId)
```

Comment:
nit: might be worth writing e.message or something to better understand what caused the failure.

To be honest, I think one Error: "Failed to process message" could also work for both JSON.parse and Zod.parse, since error message would be the same otherwise.